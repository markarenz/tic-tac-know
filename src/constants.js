// STRINGS
const PLAYER = 'player';
const CPU = 'cpu';
const X = 'x';
const O = 'o';

const AI_LEVELS = {
  AI_LEVEL_RND: 'ai_level_rnd', // Always random
  AI_LEVEL_KI: 'ai_level_ki', // Killer instinct, random until a winning move is found
  AI_LEVEL_HAH: 'ai_level_hah', // Half and half, sometimes SMRT and sometimes KI
  AI_LEVEL_SMRT: 'ai_level_smrt', // Use past games to learn how to win or draw
};

const symbols = [X, O];
const sides = [PLAYER, CPU];
const coinSides = ['heads', 'tails'];
const initOpenCells = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const winningPatterns = [
  [1, 2, 3], // TOP ROW
  [4, 5, 6], // MIDDLE ROW
  [7, 8, 9], // BOTTOM ROW
  [1, 4, 7], // LEFT COL
  [2, 5, 8], // MIDDLE COL
  [3, 6, 9], // RIGHT COL
  [1, 5, 9], // TL-BR DIAG
  [3, 5, 7], // TR-BL DIAG
];

const LOCAL_STORAGE_KEYS = {
  HISTORY: 'tic-tac-know-history',
  WINS_LOSSES: 'tic-tac-know-wins-losses',
};

const whoseTurnMessages = {
  player: 'Your turn',
  cpu: 'CPU Turn',
};
const gameOverMessages = {
  player: 'YOU WIN!',
  cpu: 'CPU WINS',
  draw: 'The Game is a draw.',
};

const APP_STATES = {
  GAME: 'game',
  MENU: 'menu',
};

const MENU_STATUSES = {
  PICK_SIDE: 'pick-side',
  PICK_AI: 'pick-ai',
  START: 'start',
  COMPLETE: 'complete',
};

const sideMessages = {
  x: 'You are playing as X. The CPU is playing as O.',
  o: 'You are playing as O. The CPU is playing as X.',
};
const aiLevelLabels = {
  ai_level_rnd: 'Herp Derp',
  ai_level_ki: 'Killer Instinct',
  ai_level_hah: 'Half & Half',
  ai_level_smrt: 'Mastermind',
};
const aiLevelMessages = {
  ai_level_rnd:
    'Your oppnent is, to put it bluntly, not terribly bright, making moves (and inapropriate noises) at random.',
  ai_level_ki:
    'Your opponent knows enough to be annoyingly good, but just like all your friends from college, it cannot plan ahead.',
  ai_level_hah:
    'Your opponent is playing with one chip tied behind its back, using a mix of strategies, sometimes smart and sometimes not.',
  ai_level_smrt:
    "Your opponent is smarter than you are, more disciplined and, let's face it, better looking. You don't stand a chance.",
};

//
export {
  symbols,
  sides,
  coinSides,
  PLAYER,
  CPU,
  X,
  O,
  whoseTurnMessages,
  gameOverMessages,
  sideMessages,
  aiLevelMessages,
  aiLevelLabels,
  initOpenCells,
  winningPatterns,
  AI_LEVELS,
  LOCAL_STORAGE_KEYS,
  APP_STATES,
  MENU_STATUSES,
};
