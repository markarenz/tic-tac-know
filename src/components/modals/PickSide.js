import React from 'react';
import PropTypes from 'prop-types';
import { TransitionWrap, MenuNav } from '../../components';
import { X, O, sideMessages } from '../../constants';

const PickSide = ({ show, playerSide, handleSelectSide, handleClickNext, handleClickPrev }) => {
  const btnXOBaseClass =
    'text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1]';
  const btnClasses = {
    selected:
      'bg-gradient-to-t from-yellow-900 to-yellow-400 rounded-lg border-purple-400 text-white',
    unselected: 'bg-gray-400 border-gray-800 rounded-lg relative text-gray-200',
  };

  return (
    <TransitionWrap show={show} className="">
      <div className="bg-purple-900 w-screen border-2 border-purple-400 rounded-lg p-4 w-full max-w-xl drop-shadow-xl">
        <div id="menu-pick-side" className="text-2xl uppercase text-white">
          Pick a side:
        </div>
        <div className="grid grid-cols-2 gap-4 mt-4">
          <div className="text-center">
            <button
              data-testid={`test-btn-side-1`}
              onClick={() => handleSelectSide(X)}
              className={`${btnXOBaseClass} ${
                playerSide === X ? btnClasses.selected : btnClasses.unselected
              }`}
            >
              X
            </button>
          </div>
          <div className="text-center">
            <button
              data-testid={`test-btn-side-2`}
              onClick={() => handleSelectSide(O)}
              className={`${btnXOBaseClass} ${
                playerSide === O ? btnClasses.selected : btnClasses.unselected
              }`}
            >
              O
            </button>
          </div>
        </div>
        <div>
          <div className="pt-4 text-white italic text-xl leading-tight">
            {sideMessages[playerSide]}
          </div>
        </div>
        <MenuNav
          handleClickPrev={handleClickPrev}
          handleClickNext={handleClickNext}
          mode="pickSide"
        />
      </div>
    </TransitionWrap>
  );
};

PickSide.propTypes = {
  show: PropTypes.bool.isRequired,
  playerSide: PropTypes.string.isRequired,
  handleSelectSide: PropTypes.func.isRequired,
  handleClickNext: PropTypes.func.isRequired,
  handleClickPrev: PropTypes.func.isRequired,
};
export default PickSide;
