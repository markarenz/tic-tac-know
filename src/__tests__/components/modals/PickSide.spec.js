import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import PickSide from '../../../components/modals/PickSide';
import { AI_LEVELS, X, O } from '../../../constants';

const mocks = {
  show: true,
  playerSide: X,
  handleSelectSide: jest.fn(),
  handleClickPrev: jest.fn(),
  handleClickNext: jest.fn(),
};
describe('PickAILevel - X', () => {
  it('renders component', async () => {
    const container = render(
      <PickSide
        show={mocks.show}
        playerSide={mocks.playerSide}
        handleSelectSide={mocks.handleSelectSide}
        handleClickPrev={mocks.handleClickPrev}
        handleClickNext={mocks.handleClickNext}
      />
    );
    await act(async () => {
      fireEvent(
        screen.getByTestId('test-btn-side-2'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
    });
    await new Promise((resolve) => setTimeout(resolve, 0));
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
  it('renders component - O', async () => {
    const container = render(
      <PickSide
        show={mocks.show}
        playerSide={O}
        handleSelectSide={mocks.handleSelectSide}
        handleClickPrev={mocks.handleClickPrev}
        handleClickNext={mocks.handleClickNext}
      />
    );
    await act(async () => {
      fireEvent(
        screen.getByTestId('test-btn-side-1'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
    });
    await new Promise((resolve) => setTimeout(resolve, 0));
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
