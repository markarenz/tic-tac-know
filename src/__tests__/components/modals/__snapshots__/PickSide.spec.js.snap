// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickAILevel - X renders component - O 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=""
        data-testid="transition-wrap"
      >
        <div
          class="bg-purple-900 w-screen border-2 border-purple-400 rounded-lg p-4 w-full max-w-xl drop-shadow-xl"
        >
          <div
            class="text-2xl uppercase text-white"
            id="menu-pick-side"
          >
            Pick a side:
          </div>
          <div
            class="grid grid-cols-2 gap-4 mt-4"
          >
            <div
              class="text-center"
            >
              <button
                class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gray-400 border-gray-800 rounded-lg relative text-gray-200"
                data-testid="test-btn-side-1"
              >
                X
              </button>
            </div>
            <div
              class="text-center"
            >
              <button
                class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gradient-to-t from-yellow-900 to-yellow-400 rounded-lg border-purple-400 text-white"
                data-testid="test-btn-side-2"
              >
                O
              </button>
            </div>
          </div>
          <div>
            <div
              class="pt-4 text-white italic text-xl leading-tight"
            >
              You are playing as O. The CPU is playing as X.
            </div>
          </div>
          <div
            class="grid grid-cols-2 mt-4 text-white"
          >
            <div
              class="text-left"
            >
              <button
                class="text-4xl hover:scale-125 transition-scale duration-300"
                data-testid="menu-nav-prev"
              >
                ◀
              </button>
            </div>
            <div
              class="text-right"
            >
              <button
                class="text-4xl hover:scale-125 transition-scale duration-300"
                data-testid="menu-nav-next"
              >
                ▶
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=""
      data-testid="transition-wrap"
    >
      <div
        class="bg-purple-900 w-screen border-2 border-purple-400 rounded-lg p-4 w-full max-w-xl drop-shadow-xl"
      >
        <div
          class="text-2xl uppercase text-white"
          id="menu-pick-side"
        >
          Pick a side:
        </div>
        <div
          class="grid grid-cols-2 gap-4 mt-4"
        >
          <div
            class="text-center"
          >
            <button
              class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gray-400 border-gray-800 rounded-lg relative text-gray-200"
              data-testid="test-btn-side-1"
            >
              X
            </button>
          </div>
          <div
            class="text-center"
          >
            <button
              class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gradient-to-t from-yellow-900 to-yellow-400 rounded-lg border-purple-400 text-white"
              data-testid="test-btn-side-2"
            >
              O
            </button>
          </div>
        </div>
        <div>
          <div
            class="pt-4 text-white italic text-xl leading-tight"
          >
            You are playing as O. The CPU is playing as X.
          </div>
        </div>
        <div
          class="grid grid-cols-2 mt-4 text-white"
        >
          <div
            class="text-left"
          >
            <button
              class="text-4xl hover:scale-125 transition-scale duration-300"
              data-testid="menu-nav-prev"
            >
              ◀
            </button>
          </div>
          <div
            class="text-right"
          >
            <button
              class="text-4xl hover:scale-125 transition-scale duration-300"
              data-testid="menu-nav-next"
            >
              ▶
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`PickAILevel - X renders component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class=""
        data-testid="transition-wrap"
      >
        <div
          class="bg-purple-900 w-screen border-2 border-purple-400 rounded-lg p-4 w-full max-w-xl drop-shadow-xl"
        >
          <div
            class="text-2xl uppercase text-white"
            id="menu-pick-side"
          >
            Pick a side:
          </div>
          <div
            class="grid grid-cols-2 gap-4 mt-4"
          >
            <div
              class="text-center"
            >
              <button
                class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gradient-to-t from-yellow-900 to-yellow-400 rounded-lg border-purple-400 text-white"
                data-testid="test-btn-side-1"
              >
                X
              </button>
            </div>
            <div
              class="text-center"
            >
              <button
                class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gray-400 border-gray-800 rounded-lg relative text-gray-200"
                data-testid="test-btn-side-2"
              >
                O
              </button>
            </div>
          </div>
          <div>
            <div
              class="pt-4 text-white italic text-xl leading-tight"
            >
              You are playing as X. The CPU is playing as O.
            </div>
          </div>
          <div
            class="grid grid-cols-2 mt-4 text-white"
          >
            <div
              class="text-left"
            >
              <button
                class="text-4xl hover:scale-125 transition-scale duration-300"
                data-testid="menu-nav-prev"
              >
                ◀
              </button>
            </div>
            <div
              class="text-right"
            >
              <button
                class="text-4xl hover:scale-125 transition-scale duration-300"
                data-testid="menu-nav-next"
              >
                ▶
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class=""
      data-testid="transition-wrap"
    >
      <div
        class="bg-purple-900 w-screen border-2 border-purple-400 rounded-lg p-4 w-full max-w-xl drop-shadow-xl"
      >
        <div
          class="text-2xl uppercase text-white"
          id="menu-pick-side"
        >
          Pick a side:
        </div>
        <div
          class="grid grid-cols-2 gap-4 mt-4"
        >
          <div
            class="text-center"
          >
            <button
              class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gradient-to-t from-yellow-900 to-yellow-400 rounded-lg border-purple-400 text-white"
              data-testid="test-btn-side-1"
            >
              X
            </button>
          </div>
          <div
            class="text-center"
          >
            <button
              class="text-8xl relative h-full w-full py-4 border-2 transition-all duration-300 hover:scale-[1.1] bg-gray-400 border-gray-800 rounded-lg relative text-gray-200"
              data-testid="test-btn-side-2"
            >
              O
            </button>
          </div>
        </div>
        <div>
          <div
            class="pt-4 text-white italic text-xl leading-tight"
          >
            You are playing as X. The CPU is playing as O.
          </div>
        </div>
        <div
          class="grid grid-cols-2 mt-4 text-white"
        >
          <div
            class="text-left"
          >
            <button
              class="text-4xl hover:scale-125 transition-scale duration-300"
              data-testid="menu-nav-prev"
            >
              ◀
            </button>
          </div>
          <div
            class="text-right"
          >
            <button
              class="text-4xl hover:scale-125 transition-scale duration-300"
              data-testid="menu-nav-next"
            >
              ▶
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
